zerver/tests/test_events.py

```diff
@@ -2745,6 +2884,44 @@ def test_change_user_role_for_restricted_users(self) -> None:
             elif role == UserProfile.ROLE_MEMBER:
                 check_user_group_add_members("events[2]", events[2])
 
+    def test_gain_access_through_metadata_groups(self) -> None:
+        reset_email_visibility_to_everyone_in_zulip_realm()
+
+        # Important: We need to refresh from the database here so that
+        # we don't have a stale UserProfile object with an old value
+        # for email being passed into this next function.
+        self.user_profile.refresh_from_db()
+
+        private_stream_1 = self.make_stream("private_stream_1", invite_only=True)
+        # We add this subscriber to make sure that event
+        # applies correctly to the `subscribers` key of
+        # stream data of "streams".
+        self.subscribe(self.example_user("cordelia"), "private_stream_1")
+        with self.verify_action(num_events=1, include_subscribers=True) as events:
+            do_change_stream_group_based_setting(
+                private_stream_1,
+                "can_administer_channel_group",
+                UserGroupMembersDict(direct_members=[self.user_profile.id], direct_subgroups=[]),
+                acting_user=self.user_profile,
+            )
+        check_stream_create("events[0]", events[0])
+
+        private_stream_2 = self.make_stream("private_stream_2", invite_only=True)
+        # We add this subscriber to make sure that event
+        # applies correctly to the `unsubscribed` key of
+        # stream data of "streams".
+        self.subscribe(self.example_user("cordelia"), "private_stream_2")
+        self.subscribe(self.user_profile, "private_stream_2")
+        self.unsubscribe(self.user_profile, "private_stream_2")
+        with self.verify_action(num_events=1, include_subscribers=True) as events:
+            do_change_stream_group_based_setting(
+                private_stream_2,
+                "can_administer_channel_group",
+                UserGroupMembersDict(direct_members=[self.user_profile.id], direct_subgroups=[]),
+                acting_user=self.user_profile,
+            )
+        check_stream_create("events[0]", events[0])
```

Comment:
This test feels a bit like a `noop` test on reading, but without this setting `subscribers=[]` in the apply_event code on stream create would not failed tests for the unsubscribed case. 

I was thinking of figuring out a way to only calculate this when a stream is not being created but a user is just gaining access to a stream via stream create, but even if a new stream is being created, not sure why we would want to set `subscribers=[]`

Reply:
I think `subscribers=[]` makes sense if we expect there to be a separate `peer_add` event that will include the subscriber data. 

I think the actual bug here is that we have not yet integrated metadata access with the `send_peer_subscriber_events` code path, which is how `peer_add` events are usually issue when someone is subscribed to a private channel that they previously could not see.

It might still be the case that merging this is a reasonable checkpoint (in part because if the `peer_add` events are sent, they will be a noop if we already have subscriber data), but we're papering over a problem by changing `subscribers=[]`.